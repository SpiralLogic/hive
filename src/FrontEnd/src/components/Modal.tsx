import '../css/modal.css'
import { FunctionComponent, h } from 'preact';

const Modal: FunctionComponent<{ name: string; onClose: () => void }> = (props) => (
  <div class="modal">
    <div role="dialog" title={props.name} class={`${props.name}`}>
      {props.children}
      <button title="Close" onClick={props.onClose}>
        Close
      </button>
    </div>
  </div>
);

Modal.displayName = 'Modal';
export default Modal;
