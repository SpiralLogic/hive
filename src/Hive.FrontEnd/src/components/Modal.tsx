import '../css/modal.css';

import { ComponentChildren } from 'preact';
import { useEffect, useRef } from 'preact/hooks';
import { effect, Signal } from '@preact/signals';

type Properties = {
  open: Signal<boolean>;
  onClose: () => void;
  class?: string;
  children?: ComponentChildren;
  title: string;
};

const Modal = (properties: Properties) => {
  const { children, open, onClose, title } = properties;
  const reference = useRef<HTMLDialogElement>(null);

  useEffect(() => {
    const dispose = effect(() => {
      const dialogNode = reference.current;
      if (!open.value) return;
      if (!dialogNode?.open) dialogNode?.showModal();
      dialogNode?.addEventListener('close', onClose);
    });

    return () => {
      reference.current?.removeEventListener('close', onClose);
      dispose();
    };
  }, [open, onClose]);

  const closeHandler = () => {
    reference.current?.close();
  };

  return (
    <dialog ref={reference} aria-label={title} class={properties.class}>
      <button class="close" aria-label="Close Dialog" onClick={closeHandler}>
        X
      </button>
      {children}
    </dialog>
  );
};

Modal.displayName = 'Modal';
export default Modal;
