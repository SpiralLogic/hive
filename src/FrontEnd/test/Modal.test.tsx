import { render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { h } from 'preact';
import Modal from '../src/components/Modal';
import { renderElement } from './test-helpers';

describe('modal snapshot tests', () => {
  it('click on modal background calls close', () => {
    const close = jest.fn();
    render(<Modal visible={true} name="test" onClose={close} />);
    userEvent.click(screen.getByTestId('modal'));

    expect(close).toHaveBeenCalledWith();
  });

  it('snapshot', () => {
    expect(renderElement(<Modal visible={true} name="test" onClose={() => ({})} />)).toMatchSnapshot();
  });
});
