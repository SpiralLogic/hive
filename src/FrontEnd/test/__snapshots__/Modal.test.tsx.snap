// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal snapshot tests snapshot 1`] = `
<div
  class="modal"
>
  <div
    class="test"
    role="dialog"
    title="test"
  >
    <button
      class="close"
      title="Close"
    >
      X
    </button>
  </div>
</div>
`;
